<a href="https://colab.research.google.com/github/khairihr/MachineLearning/blob/main/TaskWeek9/Regression_Neural_Network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nama : Khairi Hibatullah Ridho

NIM: 1103228240

Prompt: saya punya dua array, x dan y. x berisi [1.0, 2.0, 3.0, 4.0, 5.0, 6.0], dan y berisi [4.0, 6.0, 8.0, 10.0, 12.0, 14.0]. Saya ingin tahu apa yang bisa saya lakukan dengan array ini, seperti pengolahan data atau analisis sederhana. Ada saran atau ide tentang apa yang bisa saya coba?

In [1]:
import numpy as np

x = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], dtype=float)
y = np.array([4.0, 6.0, 8.0, 10.0, 12.0, 14.0], dtype=float)

Kode tersebut adalah contoh penggunaan library NumPy di Python untuk membuat dua array NumPy, yaitu `x` dan `y`, yang berisi data numerik. Mari kita jelaskan baris per baris:

1. `import numpy as np`: Ini adalah pernyataan impor yang mengimpor library NumPy dengan alias `np`. Dengan menggunakan alias `np`, kita dapat merujuk ke fungsi dan objek NumPy dengan lebih singkat, misalnya, `np.array` daripada `numpy.array`.

2. `x = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], dtype=float)`: Baris ini membuat array NumPy `x` yang berisi angka-angka [1.0, 2.0, 3.0, 4.0, 5.0, 6.0]. Fungsi `np.array()` digunakan untuk mengubah daftar (list) Python atau urutan lainnya menjadi array NumPy. `dtype=float` digunakan untuk menetapkan tipe data elemen-elemen dalam array sebagai float. Dalam hal ini, semua elemen dalam `x` akan memiliki tipe data float.

3. `y = np.array([4.0, 6.0, 8.0, 10.0, 12.0, 14.0], dtype=float)`: Ini adalah langkah yang serupa dengan langkah sebelumnya, tetapi kali ini kita membuat array NumPy `y` dengan nilai [4.0, 6.0, 8.0, 10.0, 12.0, 14.0] dan juga menjadikan semua elemennya sebagai float.

Hasil dari kode ini adalah dua array NumPy, `x` dan `y`, yang dapat digunakan untuk melakukan berbagai operasi numerik, seperti perhitungan matematika, analisis data, atau pengolahan sinyal, tergantung pada tujuan Anda dalam pengembangan kode.

Prompt: Sekarang saya ingin mencoba membangun model jaringan saraf tiruan sederhana menggunakan TensorFlow

In [2]:
import tensorflow as tf
from tensorflow import keras

model_1_layer = tf.keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])

Kode ini adalah contoh penggunaan TensorFlow, sebuah library yang digunakan untuk pembelajaran mesin dan pengembangan model jaringan saraf tiruan (neural networks) di Python. Mari kita jelaskan baris per baris:

1. `import tensorflow as tf`: Ini adalah pernyataan impor yang mengimpor library TensorFlow dan memberi alias `tf` ke library tersebut. Ini memungkinkan Anda untuk menggunakan berbagai fungsi dan objek yang disediakan oleh TensorFlow dalam kode Anda dengan memanggil `tf`.

2. `from tensorflow import keras`: Ini adalah pernyataan impor yang mengimpor modul `keras` dari TensorFlow. Keras adalah API tingkat tinggi yang terintegrasi dengan TensorFlow, yang memudahkan pembuatan dan pelatihan model jaringan saraf tiruan dengan cara yang lebih sederhana.

3. `model_1_layer = tf.keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])`: Pada baris ini, Anda membuat model jaringan saraf tiruan sederhana menggunakan Keras API yang terintegrasi dengan TensorFlow. Model ini adalah model berlapis tunggal dengan satu lapisan Dense (juga dikenal sebagai lapisan fully connected atau lapisan densely connected).

   - `tf.keras.Sequential()`: Ini adalah fungsi untuk membuat model berurutan, yang berarti lapisan-lapisan dalam model diatur secara berurutan satu per satu.
   
   - `[keras.layers.Dense(units=1, input_shape=[1])]`: Ini adalah daftar (list) yang berisi satu lapisan Dense. Lapisan Dense ini memiliki dua parameter penting:
      - `units=1`: Menunjukkan bahwa lapisan ini memiliki satu neuron (unit) dalam lapisan tersebut. Ini adalah lapisan output dari model.
      - `input_shape=[1]`: Menunjukkan bahwa input yang masuk ke model memiliki satu fitur (atau dimensi) saja. Dalam hal ini, ini adalah model regresi linier sederhana yang mengambil satu angka sebagai input dan menghasilkan satu angka sebagai output.

Jadi, hasil dari kode ini adalah pembuatan model jaringan saraf tiruan dengan satu lapisan Dense yang sederhana, yang dapat digunakan untuk berbagai tugas seperti regresi linier sederhana atau tugas-tugas lain yang melibatkan pemrosesan data numerik. Model ini dapat dikompilasi dan dilatih dengan data yang sesuai untuk tugas yang diinginkan.

prompt: saya ingin mengompilasi model ini untuk melatihnya. Di sini saya menggunakan optimisasi Stochastic Gradient Descent (SGD) dan fungsi kerugian Mean Squared Error (MSE)

In [3]:
model_1_layer.compile(optimizer='sgd', loss='mean_squared_error')

Kode `model_1_layer.compile(optimizer='sgd', loss='mean_squared_error')` adalah langkah penting dalam persiapan model jaringan saraf tiruan untuk pelatihan. Mari kita jelaskan apa yang terjadi dalam kode ini:

1. `model_1_layer`: Ini adalah objek model jaringan saraf tiruan yang telah Anda definisikan sebelumnya. Model ini memiliki satu lapisan Dense.

2. `.compile()`: Ini adalah metode yang digunakan untuk mengkonfigurasi beberapa aspek penting dari model sebelum melatihnya. Dalam metode `.compile()`, Anda menentukan dua parameter utama:

   - `optimizer='sgd'`: Parameter ini mengacu pada algoritma optimisasi yang akan digunakan selama pelatihan model. Dalam kasus ini, Anda menggunakan Stochastic Gradient Descent (SGD) sebagai optimizer. SGD adalah salah satu algoritma optimisasi yang paling umum digunakan dalam pembelajaran mesin dan jaringan saraf tiruan. Tujuan SGD adalah untuk meminimalkan fungsi kerugian (loss function) dengan menyesuaikan bobot dan bias dalam model berdasarkan gradien fungsi kerugian.

   - `loss='mean_squared_error'`: Parameter ini mengacu pada fungsi kerugian yang akan digunakan untuk mengukur seberapa baik atau buruk model Anda bekerja selama pelatihan. Dalam kasus ini, Anda menggunakan Mean Squared Error (MSE) sebagai fungsi kerugian. MSE adalah fungsi yang sering digunakan dalam tugas regresi, dan ia menghitung rata-rata kuadrat selisih antara prediksi model dan nilai sebenarnya. Tujuannya adalah untuk meminimalkan selisih antara prediksi dan target sebanyak mungkin.

Setelah menjalankan kode ini, model Anda siap untuk melatih data dengan menggunakan SGD sebagai algoritma optimisasi dan mengukur kinerjanya menggunakan MSE sebagai fungsi kerugian. Anda dapat menggunakan metode `.fit()` untuk melatih model Anda dengan data yang sesuai setelah pengaturan ini selesai.

In [4]:
model_1_layer.fit(x, y, epochs=50)

Epoch 1/50
1/1 [==============================] - 0s 309ms/step - loss: 202.4893
Epoch 2/50
1/1 [==============================] - 0s 13ms/step - loss: 93.8628
Epoch 3/50
1/1 [==============================] - 0s 7ms/step - loss: 43.5861
Epoch 4/50
1/1 [==============================] - 0s 8ms/step - loss: 20.3155
Epoch 5/50
1/1 [==============================] - 0s 9ms/step - loss: 9.5441
Epoch 6/50
1/1 [==============================] - 0s 9ms/step - loss: 4.5578
Epoch 7/50
1/1 [==============================] - 0s 5ms/step - loss: 2.2489
Epoch 8/50
1/1 [==============================] - 0s 6ms/step - loss: 1.1793
Epoch 9/50
1/1 [==============================] - 0s 4ms/step - loss: 0.6832
Epoch 10/50
1/1 [==============================] - 0s 6ms/step - loss: 0.4526
Epoch 11/50
1/1 [==============================] - 0s 4ms/step - loss: 0.3449
Epoch 12/50
1/1 [==============================] - 0s 6ms/step - loss: 0.2941
Epoch 13/50
1/1 [==============================] - 0s 5ms/step - 


- `Epoch 1/50`, `Epoch 2/50`, ..., `Epoch 50/50`: Ini adalah tampilan dari setiap epoch selama pelatihan. Pada setiap epoch, model Anda melakukan satu kali iterasi melalui seluruh dataset pelatihan untuk memperbarui bobot dan biasnya. Dalam contoh Anda, pelatihan dilakukan selama 50 epoch.

- `1/1 [==============================]`: Ini mengindikasikan bahwa pada setiap epoch, hanya ada satu batch data yang digunakan. Jumlah batch dapat bervariasi tergantung pada bagaimana Anda mengatur pelatihan Anda dan ukuran dataset.

- `- 0s XXms/step`: Ini menunjukkan waktu yang diperlukan untuk menyelesaikan satu epoch pelatihan. Waktu ini dapat bervariasi tergantung pada kompleksitas model Anda dan kecepatan komputasi Anda.

- `loss: 202.4893`, `loss: 93.8628`, ..., `loss: 0.1908`: Ini adalah nilai fungsi kerugian (loss) pada setiap epoch. Tujuan dari pelatihan adalah untuk meminimalkan nilai fungsi kerugian ini. Pada awalnya, nilai loss mungkin cukup tinggi, tetapi seiring berjalannya pelatihan, model Anda belajar dan loss akan cenderung berkurang. Semakin rendah nilai loss pada akhir pelatihan, semakin baik model Anda dalam melakukan prediksi.

- `<keras.src.callbacks.History at 0x78e80b68f8b0>`: Ini adalah keluaran dari pelatihan model dan menunjukkan bahwa pelatihan telah selesai. Objek `History` ini biasanya digunakan untuk memvisualisasikan metrik pelatihan, seperti loss, menggunakan grafik atau plot.

Dengan melihat output ini, Anda dapat memantau kemajuan pelatihan model Anda. Tujuannya adalah melihat penurunan nilai loss dari epoch ke epoch, yang menunjukkan bahwa model Anda belajar dan semakin baik dalam melakukan tugas yang ditugaskan.

In [5]:
print(model_1_layer.predict([10.0]))

1/1 [==============================] - 0s 62ms/step
[[23.325106]]


In [6]:
model_1_layer.fit(x, y, epochs=100)

Epoch 1/100
1/1 [==============================] - 0s 8ms/step - loss: 0.1894
Epoch 2/100
1/1 [==============================] - 0s 5ms/step - loss: 0.1880
Epoch 3/100
1/1 [==============================] - 0s 7ms/step - loss: 0.1866
Epoch 4/100
1/1 [==============================] - 0s 15ms/step - loss: 0.1853
Epoch 5/100
1/1 [==============================] - 0s 6ms/step - loss: 0.1839
Epoch 6/100
1/1 [==============================] - 0s 7ms/step - loss: 0.1826
Epoch 7/100
1/1 [==============================] - 0s 7ms/step - loss: 0.1812
Epoch 8/100
1/1 [==============================] - 0s 6ms/step - loss: 0.1799
Epoch 9/100
1/1 [==============================] - 0s 8ms/step - loss: 0.1786
Epoch 10/100
1/1 [==============================] - 0s 6ms/step - loss: 0.1773
Epoch 11/100
1/1 [==============================] - 0s 6ms/step - loss: 0.1760
Epoch 12/100
1/1 [==============================] - 0s 6ms/step - loss: 0.1747
Epoch 13/100
1/1 [==============================] - 0s 12ms/

In [7]:
print(model_1_layer.predict([10.0]))

1/1 [==============================] - 0s 26ms/step
[[22.919321]]


In [8]:
model_1_layer.fit(x, y, epochs=150)

Epoch 1/150
1/1 [==============================] - 0s 6ms/step - loss: 0.0911
Epoch 2/150
1/1 [==============================] - 0s 5ms/step - loss: 0.0905
Epoch 3/150
1/1 [==============================] - 0s 7ms/step - loss: 0.0898
Epoch 4/150
1/1 [==============================] - 0s 7ms/step - loss: 0.0892
Epoch 5/150
1/1 [==============================] - 0s 6ms/step - loss: 0.0885
Epoch 6/150
1/1 [==============================] - 0s 5ms/step - loss: 0.0879
Epoch 7/150
1/1 [==============================] - 0s 6ms/step - loss: 0.0872
Epoch 8/150
1/1 [==============================] - 0s 6ms/step - loss: 0.0866
Epoch 9/150
1/1 [==============================] - 0s 8ms/step - loss: 0.0860
Epoch 10/150
1/1 [==============================] - 0s 7ms/step - loss: 0.0853
Epoch 11/150
1/1 [==============================] - 0s 7ms/step - loss: 0.0847
Epoch 12/150
1/1 [==============================] - 0s 6ms/step - loss: 0.0841
Epoch 13/150
1/1 [==============================] - 0s 6ms/st

In [9]:
print(model_1_layer.predict([10.0]))

1/1 [==============================] - 0s 40ms/step
[[22.531242]]


In [10]:
model_1_layer.fit(x, y, epochs=200)

Epoch 1/200
1/1 [==============================] - 0s 10ms/step - loss: 0.0304
Epoch 2/200
1/1 [==============================] - 0s 7ms/step - loss: 0.0302
Epoch 3/200
1/1 [==============================] - 0s 6ms/step - loss: 0.0300
Epoch 4/200
1/1 [==============================] - 0s 6ms/step - loss: 0.0298
Epoch 5/200
1/1 [==============================] - 0s 7ms/step - loss: 0.0296
Epoch 6/200
1/1 [==============================] - 0s 6ms/step - loss: 0.0293
Epoch 7/200
1/1 [==============================] - 0s 6ms/step - loss: 0.0291
Epoch 8/200
1/1 [==============================] - 0s 5ms/step - loss: 0.0289
Epoch 9/200
1/1 [==============================] - 0s 5ms/step - loss: 0.0287
Epoch 10/200
1/1 [==============================] - 0s 5ms/step - loss: 0.0285
Epoch 11/200
1/1 [==============================] - 0s 7ms/step - loss: 0.0283
Epoch 12/200
1/1 [==============================] - 0s 6ms/step - loss: 0.0281
Epoch 13/200
1/1 [==============================] - 0s 13ms/

In [11]:
print(model_1_layer.predict([10.0]))

1/1 [==============================] - 0s 84ms/step
[[22.2557]]


In [12]:
model_1_layer.fit(x, y, epochs=200)

Epoch 1/200
1/1 [==============================] - 0s 36ms/step - loss: 0.0071
Epoch 2/200
1/1 [==============================] - 0s 12ms/step - loss: 0.0070
Epoch 3/200
1/1 [==============================] - 0s 9ms/step - loss: 0.0069
Epoch 4/200
1/1 [==============================] - 0s 13ms/step - loss: 0.0069
Epoch 5/200
1/1 [==============================] - 0s 17ms/step - loss: 0.0068
Epoch 6/200
1/1 [==============================] - 0s 13ms/step - loss: 0.0068
Epoch 7/200
1/1 [==============================] - 0s 11ms/step - loss: 0.0067
Epoch 8/200
1/1 [==============================] - 0s 12ms/step - loss: 0.0067
Epoch 9/200
1/1 [==============================] - 0s 13ms/step - loss: 0.0067
Epoch 10/200
1/1 [==============================] - 0s 11ms/step - loss: 0.0066
Epoch 11/200
1/1 [==============================] - 0s 15ms/step - loss: 0.0066
Epoch 12/200
1/1 [==============================] - 0s 23ms/step - loss: 0.0065
Epoch 13/200
1/1 [==============================] 

In [13]:
print(model_1_layer.predict([10.0]))

1/1 [==============================] - 0s 25ms/step
[[22.123072]]


In [14]:
model_3_layers = tf.keras.Sequential([
    keras.layers.Dense(units=1, input_shape=[1]),
    keras.layers.Dense(units=8),
    keras.layers.Dense(units=1)
])

In [15]:
model_3_layers.compile(optimizer='sgd', loss='mean_squared_error')

In [16]:
model_3_layers.fit(x, y, epochs=150)

Epoch 1/150
1/1 [==============================] - 0s 296ms/step - loss: 118.6007
Epoch 2/150
1/1 [==============================] - 0s 8ms/step - loss: 50.6331
Epoch 3/150
1/1 [==============================] - 0s 6ms/step - loss: 161.4918
Epoch 4/150
1/1 [==============================] - 0s 7ms/step - loss: 85.9992
Epoch 5/150
1/1 [==============================] - 0s 6ms/step - loss: 78.1621
Epoch 6/150
1/1 [==============================] - 0s 5ms/step - loss: 65.7225
Epoch 7/150
1/1 [==============================] - 0s 6ms/step - loss: 38.7650
Epoch 8/150
1/1 [==============================] - 0s 7ms/step - loss: 0.6478
Epoch 9/150
1/1 [==============================] - 0s 7ms/step - loss: 1.7201
Epoch 10/150
1/1 [==============================] - 0s 9ms/step - loss: 5.9634
Epoch 11/150
1/1 [==============================] - 0s 10ms/step - loss: 9.0818
Epoch 12/150
1/1 [==============================] - 0s 10ms/step - loss: 30.2437
Epoch 13/150
1/1 [=============================

In [17]:
print(model_3_layers.predict([10.0]))

1/1 [==============================] - 0s 50ms/step
[[24.878183]]
